src/core/image_resizer.js

```diff
@@ -0,0 +1,320 @@
+/* Copyright 2023 Mozilla Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { FeatureTest, ImageKind, shadow } from "../shared/util.js";
+
+const MIN_IMAGE_DIM = 2048;
+const MAX_IMAGE_DIM = 65536;
+
+// Large images are encoded in using the BMP format (it's a way faster than
+// encoding in PNG because there are no checksums to compute).
+// Unfortunately, createImageBitmap will create a task in the main thread in
+// order to build the bitmap, hence the main thread is blocked during the
+// decoding which can be a bit long (see bug 1817244).
```

Comment:
Blocking the main-thread seems less than ideal, any chance that we can get https://bugzilla.mozilla.org/show_bug.cgi?id=1817244 fixed (or at least assigned) before moving forward with this PR?